crates/data_components/src/arrow/write.rs

```diff
@@ -263,10 +255,14 @@ impl DataSink for MemSink {
             i = (i + 1) % num_partitions;
         }
 
-        // write the outputs into the batches
-        for (target, mut batches) in self.batches.iter().zip(new_batches.into_iter()) {
-            // Append all the new batches in one go to minimize locking overhead
-            target.write().await.append(&mut batches);
+        let mut writable_targets: Vec<_> =
```

Comment:
We currently use single partition(`self.batches`) only for data acceleration, but locking on partition level could potentially lead to issues when some partition have already been updated and some not (invalid state during refresh when we do query the same time). To prevent such cases PR locks all partitions first before writing/updating them. Note: right now there is no difference as we have single partition.

```rust
 async fn scan(
        &self,
        _state: &SessionState,
        projection: Option<&Vec<usize>>,
        _filters: &[Expr],
        _limit: Option<usize>,
    ) -> Result<Arc<dyn ExecutionPlan>> {
        let mut partitions = vec![];
        for arc_inner_vec in &self.batches {
            let inner_vec = arc_inner_vec.read().await;
            partitions.push(inner_vec.clone());
        }
        Ok(Arc::new(MemoryExec::try_new(
            &partitions,
            self.schema(),
            projection.cloned(),
        )?))
    }
```rust